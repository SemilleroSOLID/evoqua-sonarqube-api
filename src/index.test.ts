require('dotenv').config();

import fetch, { Headers } from 'node-fetch';

import SonarQubeMetricHistoryApi from "./index";

// We're patching the global fetch method to store cookies needed for
// authorizing API requests:
const cookies = new Set<string>();
globalThis.fetch = async (url: RequestInfo, init: RequestInit | undefined) => {
  if (cookies.size > 0) {
    const headers = new Headers(init?.headers);
    headers.set('Cookie', Array.from(cookies).join('; '));
    init = Object.assign(init ?? {}, { headers });
  }
  const response = await fetch(url as any, init as any);
  response.headers.raw()['set-cookie']?.forEach(cookie => cookies.add(cookie));
  return response as any;
};

describe('SonarQubeMetricHistoryApi tests', () => {
  const api = new SonarQubeMetricHistoryApi(process.env.SONAR_URL!);

  beforeAll(async () => {
    await api.login(process.env.SONAR_USERNAME!, process.env.SONAR_PASSWORD!);
  });

  it('getProjectIds', async () => {
    const projectIds = await api.getProjectIds();
    expect(projectIds).toMatchSnapshot();
  });

  it('getMetrics', async () => {
    const metrics = await api.getMetrics('SCAPE-app');
    expect(metrics).toMatchSnapshot();
  });

  it('getMetricHistory', async () => {
    const metricHistory =
      await api.getMetricHistory('SCAPE-app', 'code_smells');
    expect(metricHistory).toMatchSnapshot();
  });
});
