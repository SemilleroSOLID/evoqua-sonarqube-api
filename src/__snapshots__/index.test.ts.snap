// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SonarQubeMetricHistoryApi tests getMetricHistory 1`] = `
Object {
  "name": "code_smells",
  "values": Array [
    14,
    2,
    0,
  ],
  "versions": Array [
    "2021-05-07T09:42:24-0500",
    "2021-05-09T00:30:02-0500",
    "2021-07-29T23:39:01-0500",
  ],
}
`;

exports[`SonarQubeMetricHistoryApi tests getMetrics 1`] = `
Array [
  Object {
    "key": "new_technical_debt",
    "label": "Added Technical Debt",
  },
  Object {
    "key": "code_smells",
    "label": "Code Smells",
  },
  Object {
    "key": "effort_to_reach_maintainability_rating_a",
    "label": "Effort to Reach Maintainability Rating A",
  },
  Object {
    "key": "sqale_rating",
    "label": "Maintainability Rating",
  },
  Object {
    "key": "new_maintainability_rating",
    "label": "Maintainability Rating on New Code",
  },
  Object {
    "key": "new_code_smells",
    "label": "New Code Smells",
  },
  Object {
    "key": "sqale_index",
    "label": "Technical Debt",
  },
  Object {
    "key": "sqale_debt_ratio",
    "label": "Technical Debt Ratio",
  },
  Object {
    "key": "new_sqale_debt_ratio",
    "label": "Technical Debt Ratio on New Code",
  },
]
`;

exports[`SonarQubeMetricHistoryApi tests getProjectIds 1`] = `
Array [
  "CvLAC-scraper",
  "LogisticApp",
  "SCAPE-app",
  "TallerPSP01",
  "jlopezg8:TallerPSP1",
]
`;
